firefox-ios/EcosiaTests/Account/Auth/AuthNotificationSystemTests.swift

```diff
@@ -0,0 +1,460 @@
+            } else {
+                XCTFail("Should include auth state in userInfo")
+            }
+        } else {
+            XCTFail("Should include userInfo in notification")
+        }
+    }
+
+    // MARK: - Multi-Window Notification Tests
+
+    func testEcosiaAuthStateChanged_withMultipleWindows_postsNotificationForEachWindow() {
+        // Arrange
+        let windowUUID2 = WindowUUID()
+        let windowUUID3 = WindowUUID()
+        
+        windowRegistry.registerWindow(windowUUID2)
+        windowRegistry.registerWindow(windowUUID3)
+        
+        notificationObserver.expectNotification(
+            name: .EcosiaAuthStateChanged,
+            object: authStateManager,
+            expectedCount: 3
+        )
+
+        // Act
+        authStateManager.dispatchAuthState(isLoggedIn: true, actionType: .userLoggedIn)
+
+        // Assert
+        notificationObserver.waitForExpectations(timeout: 1.0)
+        
+        XCTAssertEqual(notificationObserver.receivedNotifications.count, 3, "Should receive 3 notifications")
+        
+        // Verify all notifications have correct content
+        let windowUUIDs = Set([testWindowUUID, windowUUID2, windowUUID3])
+        var receivedWindowUUIDs = Set<WindowUUID>()
+        
+        for notification in notificationObserver.receivedNotifications {
+            XCTAssertEqual(notification.name, .EcosiaAuthStateChanged, "Should have correct notification name")
+            
+            if let userInfo = notification.userInfo,
+               let windowUUID = userInfo["windowUUID"] as? WindowUUID {
+                receivedWindowUUIDs.insert(windowUUID)
+                XCTAssertEqual(userInfo["actionType"] as? String, "userLoggedIn", "Should have correct action type")
+            } else {
+                XCTFail("Should include window UUID in userInfo")
+            }
+        }
+        
+        XCTAssertEqual(receivedWindowUUIDs, windowUUIDs, "Should receive notifications for all windows")
+    }
+
+    // MARK: - Notification Subscription Tests
+
+    func testSubscribe_withObserver_receivesNotifications() {
+        // Arrange
+        let observer = TestNotificationObserver()
+        authStateManager.subscribe(observer: observer, selector: #selector(TestNotificationObserver.handleAuthStateChanged(_:)))
+        
+        let action = AuthStateAction(
+            type: .userLoggedIn,
+            windowUUID: testWindowUUID,
+            isLoggedIn: true
+        )
+
+        // Act
+        authStateManager.dispatch(action: action, for: testWindowUUID)
+
+        // Assert
+        // Give some time for the notification to be processed
+        let expectation = expectation(description: "Observer should receive notification")
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 1.0)
```

Comment:
Couldn't we attach the expectation completion to `handleAuthStateChanged` so no manual time needs to be added here?

Reply:
Really nice one :) Will do!